Add support for remote tier management

poornas opened the pull request:
With this change MinIO's ILM supports transitioning objects to a remote tier.
This change includes support for Azure Blob Storage, AWS S3 and Google Cloud
Storage as remote tier storage backends.

Co-authored-by: Poorna Krishnamoorthy <poorna@minio.io>
Co-authored-by: Krishna Srinivas <krishna@minio.io>
Co-authored-by: Krishnan Parthasarathi <kp@minio.io>

## Description


## Motivation and Context
Allows tiering older content to public cloud/ another MinIO cluster

## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

krishnasrinivas committed:
Add support for remote tier management

With this change MinIO's ILM supports transitioning objects to a remote tier.
This change includes support for Azure Blob Storage, AWS S3 and Google Cloud
Storage as remote tier storage backends.

Co-authored-by: Poorna Krishnamoorthy <poorna@minio.io>
Co-authored-by: Krishna Srinivas <krishna@minio.io>
Co-authored-by: Krishnan Parthasarathi <kp@minio.io>

krisis commented on cmd/api-headers.go:
```diff
@@ -197,8 +197,26 @@ func setObjectHeaders(w http.ResponseWriter, objInfo ObjectInfo, rs *HTTPRangeSp
 		if objInfo.TransitionStatus == lifecycle.TransitionComplete {
-			w.Header()[xhttp.AmzStorageClass] = []string{objInfo.StorageClass}
+			restoreHdr, ok := objInfo.UserDefined[xhttp.AmzRestore]
+			if !ok || !strings.HasPrefix(restoreHdr, "ongoing-request=false") || (!objInfo.RestoreExpires.IsZero() && time.Now().After(objInfo.RestoreExpires)) {
```
This could use a comment like,
// Check if object is being restored. For more information on `x-amz-restore` header see https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html#API_HeadObject_ResponseSyntax

poornas replied:
fixed

krisis commented on cmd/bucket-replication.go:
```diff
@@ -648,7 +648,7 @@ func replicateObject(ctx context.Context, objInfo ObjectInfo, objectAPI ObjectLa
 		if _, err = c.CopyObject(ctx, dest.Bucket, object, dest.Bucket, object, getCopyObjMetadata(objInfo, dest), dstOpts); err != nil {
 			replicationStatus = replication.Failed
-			logger.LogIf(ctx, fmt.Errorf("Unable to replicate metadata for object %s/%s(%s): %w", bucket, objInfo.Name, objInfo.VersionID, err))
+			logger.LogIf(ctx, fmt.Errorf("Unable to replicate metadata for object %s/%s(%s): %s", bucket, objInfo.Name, objInfo.VersionID, err))
```
@poornas are these actual improvements to error logged or coming from rebase conflict resolution?

poornas replied:
outdated - fixed by rebase

krisis commented on cmd/tier.go:
```diff
@@ -0,0 +1,402 @@
+/*
+ * MinIO Cloud Storage, (C) 2018-2019 MinIO, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2021 MinIO, Inc.
```

krisis commented on cmd/tier-handlers.go:
```diff
@@ -0,0 +1,152 @@
+package cmd
```
Missing copyright headers

krisis commented on cmd/warm-backend-azure.go:
```diff
@@ -0,0 +1,171 @@
+package cmd
```
missing copyright headers

krisis commented on cmd/warm-backend-gcs.go:
```diff
@@ -0,0 +1,186 @@
+package cmd
```
missing copyright headers

krisis commented on cmd/warm-backend-s3.go:
```diff
@@ -0,0 +1,111 @@
+package cmd
```
missing copyright headers

krisis commented on cmd/warm-backend.go:
```diff
@@ -0,0 +1,34 @@
+package cmd
```
missing copyright headers

krisis commented on pkg/madmin/remote-target-commands.go:
```diff
@@ -37,12 +37,13 @@ const (
 	ReplicationService ServiceType = "replication"
 	// ILMService specifies ilm service
-	ILMService ServiceType = "ilm"
+//	ILMService ServiceType = "ilm"
```
```suggestion

```

poornas replied:
removed this

krisis commented on pkg/madmin/remote-target-commands.go:
```diff
@@ -37,12 +37,13 @@ const (
 // IsValid returns true if ARN type represents replication or ilm
 func (t ServiceType) IsValid() bool {
-	return t == ReplicationService || t == ILMService
+	//	return t == ReplicationService || t == ILMService
```
```suggestion

```

krishnasrinivas committed:
Support healing of transitioned objects (#12)

Commit:
Add DESIGN.md to document ilm tiering

krisis committed:
Use journalling to mark deletion of objects in remote tier (#11)

- Also simplify helper function to cleanup transitioned objects on expiry or when
they are 'overwritten'
- Use expireTransitionedObjects in delete-object, multi-delete-objects and
put-object API handlers
